custom/nic_compliance/utils.py

```diff
@@ -0,0 +1,119 @@
+import re
+import base64
+from datetime import timedelta
+from django.conf import settings
+
+from django.contrib.auth.hashers import get_hasher
+
+from dimagi.utils.couch.cache.cache_core import get_redis_client
+
+from corehq.util.view_utils import get_request
+from custom.nic_compliance.const import EXPIRE_LOGIN_ATTEMPTS_IN, REDIS_LOGIN_ATTEMPTS_LIST_PREFIX
+
+PASSWORD_HASHER = get_hasher()
+
+
+def extract_password(obfuscated_password):
+    # Passwords set with expected padding length and format would respect this regex
+    reg_exp = r"^sha256\$([a-z0-9A-Z]{6})([a-zA-Z0-9=]*)([a-z0-9A-Z]{6})=$"
+    # match regex for padding along with inner block and find matches
+    match_result = re.match(reg_exp, obfuscated_password)
+    # ensure regex match for obfuscated password and three matches
+    if match_result and len(match_result.groups()) == 3:
+        match_groups = match_result.groups()
+        padding_left, encoded_internal_block, padding_right = match_groups
+        # b64 decode the encoded internal block to get raw internal block
+        # raw internal block = (paddling_left + encoded_password + padding_right)
+        try:
+            decoded_internal_block = base64.b64decode(encoded_internal_block)
+        except TypeError:
+            return ''
+        # match regex for padding along with b64 encoded password and find matches
+        match_result_2 = re.match(reg_exp, decoded_internal_block)
+        # ensure regex match for the internal block and 3 matches
+        if match_result_2 and len(match_result_2.groups()) == 3:
+            match_groups = match_result_2.groups()
+            # ensure the same padding was used in the internal block as the outer
+            if match_groups[0] == padding_left and match_groups[2] == padding_right:
+                b64_encoded_password = match_groups[1]
+                try:
+                    # decode to get the b64encoded real password
+                    # return password decoded for UTF-8 support
+                    return base64.b64decode(b64_encoded_password).decode('utf-8')
+                except TypeError:
+                    return ''
+            else:
+                # this sounds like someone tried to obfuscate password but failed so ignore the password submitted
+                # completely
+                return ''
+        else:
+            # this sounds like someone tried to obfuscate password but failed so ignore the password submitted
+            # completely
+            return ''
+    else:
+        # return the password received AS-IS
+        return obfuscated_password
+
+
+def hash_password(password):
+    return PASSWORD_HASHER.encode(password, PASSWORD_HASHER.salt())
+
+
+def verify_password(password, password_salt):
```

Comment:
Nitpick: the second argument is not a salt. Consider renaming to `encoded` as in https://github.com/django/django/blob/1.11/django/contrib/auth/hashers.py#L200